### PR TITLE
Generate Software Bill Of Materials (SBOMs) for our RPM packages

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR adds a few steps to the RPM Docker files:
- `Dockerfile.rpmtest` installs the generate packages on the base image
- `Dockerfile.rpmsbom` generates a SBOM in [CycloneDX](https://cyclonedx.org/) format for the installed packages

I'm skipping EL-7 packages because EL-7 will be EOL soon, and `python2-dnf` was adding a lot of complexity.

I'm attaching some generated SBOMs as an example:
[dnsdist-1.9.0-alpha0.2486.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757168/dnsdist-1.9.0-alpha0.2486.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-tools-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757169/pdns-tools-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-tinydns-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757170/pdns-backend-tinydns-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-sqlite-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757171/pdns-backend-sqlite-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-remote-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757172/pdns-backend-remote-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-postgresql-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757173/pdns-backend-postgresql-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-odbc-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757174/pdns-backend-odbc-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-lua2-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757176/pdns-backend-lua2-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-mysql-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757177/pdns-backend-mysql-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-lmdb-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757178/pdns-backend-lmdb-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-geoip-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757179/pdns-backend-geoip-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-backend-ldap-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757181/pdns-backend-ldap-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)
[pdns-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json](https://github.com/PowerDNS/pdns/files/14757182/pdns-4.8.0-beta1.2338.generatesbom.gfef1b1c63-el-9.cyclonedx.json)

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: builder-support/dockerfiles/Dockerfile.rpmtest [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# We NEED these because pdns-builder expects to find /sdist and /dist
+# in the final image
+COPY --from=sdist /sdist /sdist
+COPY --from=package-builder /dist /dist
+
+# Install built packages with dependencies
+RUN yum localinstall -y /dist/*/*.rpm
+
+# Generate SBOM
+# We should probably guard this behind "${IS_RELEASE}" = "YES"
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply builder-support/dockerfiles/Dockerfile.rpmtest:16]: It's a very good point, and the overhead is not huge either, so let's keep it enabled at all time for now!

File: builder-support/dockerfiles/Dockerfile.rpmsbom [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+RUN cd /pdns/builder-support/helpers/; \
+    if ! $(grep -q 'release 7' /etc/redhat-release); then \
+       python3 generate-sbom-dnf.py /dist/dnsdist-${BUILDER_VERSION}-${BUILDER_TARGET}.cyclonedx.json dnsdist rust.json quiche.json h2o.json; \
+    fi
+@ENDIF
+
+@IF [ -n "$M_authoritative$M_all" ]
+RUN cd /pdns/builder-support/helpers/ && \
+    if ! $(grep -q 'release 7' /etc/redhat-release); then \
+      for pkg in pdns pdns-backend-geoip pdns-backend-ldap pdns-backend-lmdb pdns-backend-lua2 pdns-backend-mysql pdns-backend-odbc pdns-backend-postgresql pdns-backend-remote pdns-backend-sqlite pdns-backend-tinydns pdns-backend-ixfrdist pdns-tools; do \
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply builder-support/dockerfiles/Dockerfile.rpmsbom:28]: I'm sure it could be! I'll look into it.

**Review by Habbie — APPROVED** [review]

File: builder-support/dockerfiles/Dockerfile.rpmsbom [hunk 3]
```diff
@@ -1,8 +1,10 @@
-# No SBOM on el-7, which only has Python 2 and will be EOL soon anyway
+ARG BUILDER_VERSION
+ARG BUILDER_TARGET
 
+# No SBOM on el-7, which only has Python 2 and will be EOL soon anyway
 RUN touch /var/lib/rpm/* && \
     if ! $(grep -q 'release 7' /etc/redhat-release); then \
-      yum upgrade -y && yum install -y python3 python3-dnf; \
```
Comment (omoerbeek) [inline builder-support/dockerfiles/Dockerfile.rpmsbom]: Was it intended to skip the upgrade?

**Review by omoerbeek — APPROVED** [review]
One question, rest looks ok.

File: builder-support/dockerfiles/Dockerfile.rpmsbom (hunk 3 above)
Comment (rgacogne) [reply builder-support/dockerfiles/Dockerfile.rpmsbom]: Yes, because otherwise `yum` might try to downgrade the installed packages!

**Pull request merged**